pkg/naming/resource_converter.go

```diff
@@ -0,0 +1,200 @@
+	}
+
+	// invert the structure for consistency with the template
+	for lbl, groupRes := range overrides {
+		infoRaw := provider.CustomMetricInfo{
+			GroupResource: schema.GroupResource{
+				Group:    groupRes.Group,
+				Resource: groupRes.Resource,
+			},
+		}
+		info, _, err := infoRaw.Normalized(converter.mapper)
+		if err != nil {
+			return nil, fmt.Errorf("unable to normalize group-resource %v: %v", groupRes, err)
+		}
+
+		converter.labelToResource[pmodel.LabelName(lbl)] = info.GroupResource
+		converter.resourceToLabel[info.GroupResource] = pmodel.LabelName(lbl)
+	}
+
+	return converter, nil
+}
+
+func (r *resourceConverter) LabelForResource(resource schema.GroupResource) (pmodel.LabelName, error) {
+	r.labelResourceMu.RLock()
+	// check if we have a cached copy or override
+	lbl, ok := r.resourceToLabel[resource]
+	r.labelResourceMu.RUnlock() // release before we call makeLabelForResource
+	if ok {
+		return lbl, nil
+	}
+
+	// NB: we don't actually care about the gap between releasing read lock
+	// and acquiring the write lock -- if we do duplicate work sometimes, so be
+	// it, as long as we're correct.
+
+	// otherwise, use the template and save the result
+	lbl, err := r.makeLabelForResource(resource)
+	if err != nil {
+		return "", fmt.Errorf("unable to convert resource %s into label: %v", resource.String(), err)
+	}
+	return lbl, nil
+}
+
+// makeLabelForResource constructs a label name for the given resource, and saves the result.
+// It must *not* be called under an existing lock.
+func (r *resourceConverter) makeLabelForResource(resource schema.GroupResource) (pmodel.LabelName, error) {
+	if r.labelTemplate == nil {
+		return "", fmt.Errorf("no generic resource label form specified for this metric")
+	}
+	buff := new(bytes.Buffer)
+
+	singularRes, err := r.mapper.ResourceSingularizer(resource.Resource)
+	if err != nil {
+		return "", fmt.Errorf("unable to singularize resource %s: %v", resource.String(), err)
+	}
+	convResource := schema.GroupResource{
+		Group:    groupNameSanitizer.Replace(resource.Group),
+		Resource: singularRes,
+	}
+
+	if err := r.labelTemplate.Execute(buff, convResource); err != nil {
+		return "", err
+	}
+	if buff.Len() == 0 {
+		return "", fmt.Errorf("empty label produced by label template")
+	}
+	lbl := pmodel.LabelName(buff.String())
+
+	r.labelResourceMu.Lock()
+	defer r.labelResourceMu.Unlock()
+
+	r.resourceToLabel[resource] = lbl
+	r.labelToResource[lbl] = resource
+	return lbl, nil
+}
+
+func (r *resourceConverter) ResourcesForSeries(series prom.Series) ([]schema.GroupResource, bool) {
+	// use an updates map to avoid having to drop the read lock to update the cache
+	// until the end.  Since we'll probably have few updates after the first run,
+	// this should mean that we rarely have to hold the write lock.
+	var resources []schema.GroupResource
+	updates := make(map[pmodel.LabelName]schema.GroupResource)
+	namespaced := false
+
+	// use an anon func to get the right defer behavior
```

Comment:
there are no returns in this anonymous function so we can just lock in the beginning and unlock afterwards

Reply:
defer avoids issues if a panic occurs